File: src/main/java/Wrapper.java
public class Wrapper {
    public String wrap(String text, int columnsWidth) {
        validate(columnsWidth);

        if (text.length() == 0) return "";

        if (text.length() <= columnsWidth) return text;

        final int wrapLimit = obtainWrapLimit(text, columnsWidth);
        final var firstWrap = text.substring(0, wrapLimit).concat("\n");
        final var remainingText = text.substring(wrapLimit).trim();
        return firstWrap.concat(wrap(remainingText, columnsWidth));
    }

    private void validate(int columnsWidth) {
        isColumnWidthZero(columnsWidth);
        isColumnWidthNegative(columnsWidth);
    }

    private void isColumnWidthZero(int columnsWidth) {
        if (columnsWidth == 0) throw new IllegalArgumentException("Zero are not allowed");
    }

    private void isColumnWidthNegative(int columnsWidth) {
        if (columnsWidth < 0) throw new IllegalArgumentException("Negative numbers are not allowed");
    }

    private int obtainWrapLimit(String text, int columnsWidth) {
        final boolean shouldWrappedInBlankSpace = isWrappedInBlankSpace(text, columnsWidth);
        return (shouldWrappedInBlankSpace) ? text.lastIndexOf(" ", columnsWidth) : columnsWidth;
    }

    private boolean isWrappedInBlankSpace(String text, int columnsWidth) {
        final int blankSpaceIndex = text.lastIndexOf(" ", columnsWidth);
        return 0 < blankSpaceIndex && blankSpaceIndex < columnsWidth;
    }
}

File: src/Wrapper.ts
export class Wrapper {
  public wrap(text: string, columnsWidth: number): string {
    this.validate(columnsWidth);

    if (text === "") return "";

    if (text.length <= columnsWidth) return text;

    const wrapLimit = this.obtainWrapLimit(text, columnsWidth);
    const [wrapped, remainingText] = this.splitFirstWrapAndRemainingText(
      text,
      wrapLimit
    );
    return wrapped + this.wrap(remainingText, columnsWidth);
  }

  private validate(columnsWidth: number) {
    this.isColumnWidthZero(columnsWidth);
    this.isColumnWidthNegative(columnsWidth);
  }

  private isColumnWidthZero(columnWidth: number) {
    if (columnWidth === 0) throw "Zero are not allowed";
  }

  private isColumnWidthNegative(columnWidth: number) {
    if (columnWidth < 0) throw "Negative numbers are not allowed";
  }

  private obtainWrapLimit(text: string, columnsWidth: number): number {
    const shouldWrappedInBlankSpace = this.isWrappedInBlankSpace(
      text,
      columnsWidth
    );
    return shouldWrappedInBlankSpace
      ? text.lastIndexOf(" ", columnsWidth)
      : columnsWidth;
  }
  private isWrappedInBlankSpace(text: string, columnsWidth: number): boolean {
    const blankSpaceIndex = text.lastIndexOf(" ", columnsWidth);
    return 0 < blankSpaceIndex && blankSpaceIndex < columnsWidth;
  }
  private splitFirstWrapAndRemainingText(
    text: string,
    wrapLimit: number
  ): string[] {
    const wrapped = text.substring(0, wrapLimit).concat("\n");
    const remainingText = text.substring(wrapLimit).trim();
    return [wrapped, remainingText];
  }
}
